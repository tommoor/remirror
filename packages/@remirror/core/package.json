{
  "name": "@remirror/core",
  "version": "1.0.0-next.56",
  "description": "Where your quest to create a world class editing experience begins.",
  "homepage": "https://github.com/remirror/remirror/tree/HEAD/packages/@remirror/core",
  "repository": "https://github.com/remirror/remirror/tree/HEAD/packages/@remirror/core",
  "license": "MIT",
  "contributors": [
    "Ifiok Jr. <ifiokotung@gmail.com>"
  ],
  "sideEffects": false,
  "exports": {
    ".": {
      "import": "./dist/remirror-core.esm.js",
      "require": "./dist/remirror-core.cjs.js",
      "browser": "./dist/remirror-core.browser.esm.js",
      "types": "./dist/remirror-core.cjs.d.ts",
      "default": "./dist/remirror-core.esm.js"
    },
    "./package.json": "./package.json",
    "./types/*": "./dist/declarations/src/*.d.ts"
  },
  "main": "dist/remirror-core.cjs.js",
  "module": "dist/remirror-core.esm.js",
  "browser": {
    "./dist/remirror-core.cjs.js": "./dist/remirror-core.browser.cjs.js",
    "./dist/remirror-core.esm.js": "./dist/remirror-core.browser.esm.js"
  },
  "types": "dist/remirror-core.cjs.d.ts",
  "files": [
    "style",
    "dist"
  ],
  "dependencies": {
    "@babel/runtime": "^7.12.5",
    "@linaria/core": "^3.0.0-beta.1",
    "@remirror/core-constants": "1.0.0-next.56",
    "@remirror/core-helpers": "1.0.0-next.56",
    "@remirror/core-types": "1.0.0-next.56",
    "@remirror/core-utils": "1.0.0-next.56",
    "@remirror/i18n": "1.0.0-next.56",
    "@remirror/icons": "0.0.0",
    "@remirror/theme": "1.0.0-next.56",
    "nanoevents": "^5.1.10",
    "tiny-warning": "^1.0.3",
    "type-fest": "^0.20.2"
  },
  "devDependencies": {
    "@remirror/pm": "1.0.0-next.56"
  },
  "peerDependencies": {
    "@remirror/pm": "1.0.0-next.56"
  },
  "publishConfig": {
    "access": "public"
  },
  "meta": {
    "sizeLimit": "40 KB"
  }
}
