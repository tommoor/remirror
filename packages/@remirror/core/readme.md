# @remirror/core

> Where your quest to create a world class editor begins.

[![Version][version]][npm] [![Weekly Downloads][downloads-badge]][npm] [![Bundled size][size-badge]][size] [![Typed Codebase][typescript]](#) [![MIT License][license]](#)

[version]: https://flat.badgen.net/npm/v/@remirror/core/next
[npm]: https://npmjs.com/package/@remirror/core/v/next
[license]: https://flat.badgen.net/badge/license/MIT/purple
[size]: https://bundlephobia.com/result?p=@remirror/core
[size-badge]: https://flat.badgen.net/bundlephobia/minzip/@remirror/core
[typescript]: https://flat.badgen.net/badge/icon/TypeScript?icon=typescript&label
[downloads-badge]: https://badgen.net/npm/dw/@remirror/core/red?icon=npm

## Installation

```bash
# yarn
yarn add @remirror/core @remirror/pm

# pnpm
pnpm add @remirror/core @remirror/pm

# npm
npm install @remirror/core @remirror/pm
```

## Usage

This is the core `remirror` package where the `Extension` and `RemirrorManager` are defined.
