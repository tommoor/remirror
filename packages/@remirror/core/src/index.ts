export * from '@remirror/core-constants';
export * from '@remirror/core-helpers';
export * from '@remirror/core-types';
export * from '@remirror/core-utils';
export type { CoreIcon } from '@remirror/icons';

export * from './builtins';

export type { DelayedValue } from './commands';
export { delayedCommand, insertText, isDelayedValue, toggleMark } from './commands';
export * from './extension';
export type {
  AddCustomHandler,
  AddHandler,
  CustomHandlerMethod,
  HandlerKeyOptions,
} from './extension/base-class';
export type {
  AttributePropFunction,
  BaseFramework,
  BaseListenerParameter,
  CreateStateFromContent,
  FocusType,
  FrameworkOutput,
  FrameworkParameter,
  FrameworkProps,
  ListenerParameter,
  PlaceholderConfig,
  RemirrorEventListener,
  RemirrorEventListenerParameter,
  RemirrorGetterParameter,
  TriggerChangeParameter,
  UpdateStateParameter,
} from './framework';
export { Framework } from './framework';
export type { AnyRemirrorManager, CreateEditorStateParameter, ManagerEvents } from './manager';
export { isRemirrorManager, RemirrorManager } from './manager';
export type {
  ApplyStateLifecycleParameter,
  BaseExtensionOptions,
  ChangedOptions,
  CommandShape,
  CreateExtensionPlugin,
  DynamicOptionsOfConstructor,
  ExcludeOptions,
  ExtensionCommandFunction,
  ExtensionCommandReturn,
  ExtensionHelperReturn,
  ExtensionStore,
  GetChangeOptionsReturn,
  GetCommands,
  GetConstructor,
  GetHelpers,
  GetNameUnion,
  GetOptions,
  OnSetOptionsParameter,
  OptionsOfConstructor,
  PickChanged,
  StateUpdateLifecycleParameter,
  UpdateReason,
  UpdateReasonParameter,
} from './types';
