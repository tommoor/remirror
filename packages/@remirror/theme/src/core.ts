import { css } from '@linaria/core';

/**
 * This is compiled into the class name `remirror-editor` and the css is
 * automatically generated and placed into the `@remirror/styles/core.css` via
 * a `linaria` build script.
 */
export const EDITOR = css`
  &.ProseMirror {
    word-wrap: break-word;
    white-space: pre-wrap;
    white-space: break-spaces;
    position: relative;
    font-variant-ligatures: none;
    font-feature-settings: 'liga' 0;

    pre {
      white-space: pre-wrap;
    }

    li {
      position: relative;
    }

    &[contenteditable='false'] {
      white-space: normal;
    }

    &[contenteditable='true'] {
      white-space: pre-wrap;
    }

    hr {
      border-color: #2e2e2e;
    }
  }

  .ProseMirror-hideselection *::selection {
    background: transparent;
  }

  .ProseMirror-hideselection *::-moz-selection {
    background: transparent;
  }

  .ProseMirror-hideselection {
    caret-color: transparent;
  }

  .ProseMirror-selectednode {
    outline: 2px solid #8cf;
  }

  /* Make sure li selections wrap around markers */

  li.ProseMirror-selectednode {
    outline: none;

    &:after {
      content: '';
      position: absolute;
      left: -32px;
      right: -2px;
      top: -2px;
      bottom: -2px;
      border: 2px solid #8cf;
      pointer-events: none;
    }
  }
` as 'rmr-editor';
