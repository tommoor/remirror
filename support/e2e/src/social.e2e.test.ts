import { getDocument, queries } from 'playwright-testing-library';
import { ElementHandle } from 'playwright-testing-library/dist/typedefs';

import { EDITOR_CLASS_SELECTOR } from '@remirror/core';

import {
  $innerHtml,
  goto,
  innerHtml,
  outerHtml,
  press,
  sel,
  skipTestOnFirefox,
  textContent,
  type,
} from './helpers';

const FIRST_PARAGRAPH_SELECTOR = `${EDITOR_CLASS_SELECTOR} > p:first-child`;

const { getByRole } = queries;
const path = __SERVER__.urls.social.empty;

describe('Social Showcase', () => {
  let $document: ElementHandle;
  let $editor: ElementHandle;

  beforeEach(async () => {
    await goto(path);
    $document = await getDocument(page);
    $editor = await getByRole($document, 'textbox');
  });

  describe('Links', () => {
    it('should have a social editor', async () => {
      await $editor.focus();
      await $editor.type('This is text https://url.com');
      await expect($innerHtml(FIRST_PARAGRAPH_SELECTOR)).resolves.toMatchSnapshot();
    });

    it('should parse simple urls', async () => {
      await $editor.type('url.com');
      await expect($innerHtml(FIRST_PARAGRAPH_SELECTOR)).resolves.toMatchSnapshot();
      await press({ key: 'Backspace' });
      await expect($innerHtml(FIRST_PARAGRAPH_SELECTOR)).resolves.toMatchSnapshot();
      await press({ key: 'Backspace' });
      await expect(innerHtml(FIRST_PARAGRAPH_SELECTOR)).resolves.toBe(`url.c`);

      await type({ text: 'o.uk' });
      await expect($innerHtml(FIRST_PARAGRAPH_SELECTOR)).resolves.toMatchSnapshot();
    });

    // TODO The 'Home' key press doesn't work on Firefox
    skipTestOnFirefox('can handle more complex interactions', async () => {
      await $editor.type('this is the first url.com');
      await press({ key: 'Enter' });
      await $editor.type('this.com is test.com');
      await press({ key: 'Home' }); // ? This does nothing on Firefox
      await type({ text: 'split.com ' });
      // eslint-disable-next-line jest/no-standalone-expect
      await expect($innerHtml(FIRST_PARAGRAPH_SELECTOR)).resolves.toMatchSnapshot();

      await press({ key: 'ArrowUp' });
      await press({ key: 'End' });
      await press({ key: 'Backspace', count: 2 });
      await type({ text: '..no .co more url please' });
      // eslint-disable-next-line jest/no-standalone-expect
      await expect(innerHtml(FIRST_PARAGRAPH_SELECTOR)).resolves.toBe(
        `this is the first url.c..no .co more url please`,
      );
    });

    it('should handle the enter key', async () => {
      await $editor.type('this is the first url.com');
      await press({ key: 'ArrowLeft', count: 3 });
      await press({ key: 'Enter' });

      await expect(innerHtml(FIRST_PARAGRAPH_SELECTOR)).resolves.toBe(`this is the first url.`);
    });

    it('should not contain false positives', async () => {
      await $editor.type('https://localhost:3000/ahttps://meowni.ca');
      await expect(innerHtml(FIRST_PARAGRAPH_SELECTOR)).resolves.toBe(
        `https://localhost:3000/ahttps://meowni.ca`,
      );
    });
  });

  describe('Mentions', () => {
    it('should not allow mixing the tags', async () => {
      await $editor.type('@#ab #@simple ');
      await expect(outerHtml(sel(EDITOR_CLASS_SELECTOR, 'a'))).rejects.toThrow();
    });

    describe('@', () => {
      it('should wrap in progress mentions in a-tag decorations', async () => {
        await $editor.type('Hello @jonathan');
        await expect(textContent(sel(EDITOR_CLASS_SELECTOR, '.suggest-at'))).resolves.toBe(
          '@jonathan',
        );
      });

      it('should not trap the arrow keys', async () => {
        await $editor.type('@a');
        await press({ key: 'ArrowLeft', count: 2 });

        await $editor.type('12 ');
        await expect($editor.innerHTML()).resolves.toMatchSnapshot();
      });

      it('should not revert the mention', async () => {
        const selector = sel(EDITOR_CLASS_SELECTOR, '.mention-at');
        await $editor.type('@a ');
        await press({ key: 'ArrowLeft' });

        await $editor.type(' ');
        await expect(textContent(selector)).resolves.toBe('@a');
      });

      it('should accept selections onEnter', async () => {
        const selector = sel(EDITOR_CLASS_SELECTOR, '.mention-at');

        await $editor.type('hello @ab');
        await press({ key: 'Enter' });
        await expect(page.$$(selector)).resolves.toHaveLength(1);
        await expect(textContent(selector)).resolves.toBe('@orangefish879'); // This might change if data changes
      });

      it('should still wrap selections when exiting without selections', async () => {
        await $editor.type('hello @ab ');
        await expect(textContent(sel(EDITOR_CLASS_SELECTOR, '.mention-at'))).resolves.toBe('@ab');
      });

      it('allows clicking on suggestions', async () => {
        const selector = '.remirror-mention-suggestions-item.highlighted';
        await $editor.type('hello @alex');

        await page.click(selector);
        await expect(textContent(sel(EDITOR_CLASS_SELECTOR, '.mention-at'))).resolves.toBe(
          '@lazymeercat594',
        );
        await expect(textContent(EDITOR_CLASS_SELECTOR)).resolves.toBe(`hello @lazymeercat594 `);
      });

      it('allows arrowing between suggesters', async () => {
        await $editor.type('hello  1');
        await press({ key: 'ArrowLeft', count: 2 });
        await $editor.type('@ab');
        await press({ key: 'ArrowRight' });
        await expect(textContent(sel(EDITOR_CLASS_SELECTOR, '.mention-at'))).resolves.toBe('@ab');
      });

      it('allows arrowing between suggesters and breaking up the suggestion', async () => {
        await $editor.type('hello  1');
        await press({ key: 'ArrowLeft' });
        await $editor.type('@ab ');
        await expect(textContent(sel(EDITOR_CLASS_SELECTOR, '.mention-at'))).resolves.toBe('@ab');
      });

      it('handles arrowing between arrowing back into mention without errors', async () => {
        await $editor.type('@ab ');
        await press({ key: 'ArrowLeft', count: 4 });
        await press({ key: 'ArrowRight' });
        await expect(textContent(sel(EDITOR_CLASS_SELECTOR, '.mention-at'))).resolves.toBe('@ab');
      });

      it('removes mark when no partial query', async () => {
        await $editor.type('@abc ');
        await press({ key: 'ArrowLeft', count: 4 });
        await $editor.type(' ');
        await expect(
          textContent(sel(EDITOR_CLASS_SELECTOR, '.mention-at')),
        ).rejects.toThrowErrorMatchingInlineSnapshot(
          `"page.$eval: Error: failed to find element matching selector \\".remirror-editor .mention-at\\""`,
        );
      });

      it('adds the mark when enter is pressed', async () => {
        const username = '@abcd1234';
        await $editor.type(username);
        await press({ key: 'Enter' });
        await expect(textContent(sel(EDITOR_CLASS_SELECTOR, '.mention-at'))).resolves.toBe(
          username,
        );
      });

      it('splits up the mark when enter is pressed', async () => {
        const username = '@abcd1234 ';
        await $editor.type(username);
        await press({ key: 'ArrowLeft', count: 3 });
        await press({ key: 'Enter' });
        await expect(textContent(sel(EDITOR_CLASS_SELECTOR, '.mention-at'))).resolves.toBe(
          '@abcd12',
        );
      });

      it('removes mentions for forward deletes', async () => {
        await $editor.type('@abc ');
        await press({ key: 'ArrowLeft', count: 5 });
        await press({ key: 'Delete' });
        await expect(
          textContent(sel(EDITOR_CLASS_SELECTOR, '.mention-at')),
        ).rejects.toThrowErrorMatchingInlineSnapshot(
          `"page.$eval: Error: failed to find element matching selector \\".remirror-editor .mention-at\\""`,
        );
      });

      it('can exit when selecting text', async () => {
        await $editor.type('@abc ');
        await press({ key: 'ArrowLeft', count: 1 });
        await page.keyboard.down('Shift');
        await press({ key: 'ArrowLeft', count: 4 });
        await page.keyboard.up('Shift');
        await press({ key: 'ArrowLeft', count: 1 });
        await $editor.type('Awesome ');

        await expect(textContent(sel(EDITOR_CLASS_SELECTOR, '.mention-at'))).resolves.toBe('@abc');
        await expect(textContent(sel(EDITOR_CLASS_SELECTOR))).resolves.toBe('Awesome @abc ');
      });

      it('only replaces the selected text in a mention', async () => {
        await $editor.type('@abc ');
        await press({ key: 'ArrowLeft', count: 2 });
        await page.keyboard.down('Shift');
        await press({ key: 'ArrowLeft', count: 3 });
        await page.keyboard.up('Shift');
        await $editor.type('d');

        await expect(textContent(sel(EDITOR_CLASS_SELECTOR))).resolves.toBe('dc ');
      });
    });

    describe('#', () => {
      it('should wrap in progress mentions in a-tag decorations', async () => {
        await $editor.type('My tag is #Topic');
        await expect(textContent(sel(EDITOR_CLASS_SELECTOR, '.suggest-tag'))).resolves.toBe(
          '#Topic',
        );
      });

      it('should accept selections onEnter', async () => {
        const selector = sel(EDITOR_CLASS_SELECTOR, '.mention-tag');

        await $editor.type('hello #T');
        await press({ key: 'Enter' });
        await expect(page.$$(selector)).resolves.toHaveLength(1);
        await expect(textContent(selector)).resolves.toBe('#Tags');
      });

      it('should still wrap selections when exiting without selections', async () => {
        await $editor.type('hello #T ');
        await expect(textContent(sel(EDITOR_CLASS_SELECTOR, '.mention-tag'))).resolves.toBe('#T');
      });

      it('allows clicking on suggesters', async () => {
        const selector = '.remirror-mention-suggestions-item.highlighted';
        await $editor.type('My #T');
        await page.click(selector);
        await expect(textContent(sel(EDITOR_CLASS_SELECTOR, '.mention-tag'))).resolves.toBe(
          '#Tags',
        );
        await expect(textContent(EDITOR_CLASS_SELECTOR)).resolves.toBe(`My #Tags `);
      });
    });
  });

  describe('Emoji', () => {
    // Emoji are being completely rewritten soon so this is temporary
    it('should be able to add emoji', async () => {
      await $editor.type('😀', { delay: 10 });
      await expect(innerHtml(sel(EDITOR_CLASS_SELECTOR, 'p'))).resolves.toBe(`😀`);
    });

    it('transforms emoticons', async () => {
      await $editor.type(':-) hello', { delay: 10 });
      await expect(innerHtml(FIRST_PARAGRAPH_SELECTOR)).resolves.toBe(`😃 hello`);
    });

    it('transforms colon emojis', async () => {
      await $editor.type(':heart:', { delay: 10 });
      await expect(innerHtml(FIRST_PARAGRAPH_SELECTOR)).resolves.toBe(`❤️`);
    });

    it('should handle multiple emoji with no spaces', async () => {
      const text = '123abcXYZ';
      await $editor.type('😀😀😀😀', { delay: 10 });
      await press({ key: 'ArrowLeft', count: 2 });
      await press({ key: 'ArrowRight' });
      await type({ text });
      await expect(textContent(FIRST_PARAGRAPH_SELECTOR)).resolves.toBe(`😀😀😀123abcXYZ😀`);
    });
  });

  describe('Combined Emoji, Mentions and AutoLinks', () => {
    it('should combine mentions emoji and links', async () => {
      await $editor.type('#awesome hello @ab 😀 google.com', { delay: 10 });
      await press({ key: 'Enter' });
      await expect($innerHtml(FIRST_PARAGRAPH_SELECTOR)).resolves.toMatchSnapshot();
    });

    it('should not replace emoji with link when no space between', async () => {
      await $editor.type('😀google.com', { delay: 10 });
      await press({ key: 'Enter' });
      await expect($innerHtml(FIRST_PARAGRAPH_SELECTOR)).resolves.toMatchSnapshot();
    });
  });
});
