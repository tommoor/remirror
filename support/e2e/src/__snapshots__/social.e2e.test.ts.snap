// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Social Showcase ExtensionUnion should combine mentions emoji and links 1`] = `
<a role="presentation"
   href="/awesome"
   class="mention mention-tag"
   data-mention-id="awesome"
   data-mention-name="tag"
>
  #awesome
</a>
hello
<a role="presentation"
   href="/ab"
   class="mention mention-at"
   data-mention-id="ab"
   data-mention-name="at"
>
  @ab
</a>
ðŸ˜€
<a href="//google.com"
   role="presentation"
>
  google.com
</a>
`;

exports[`Social Showcase ExtensionUnion should not replace emoji with link when no space between 1`] = `
ðŸ˜€
<a href="//google.com"
   role="presentation"
>
  google.com
</a>
`;

exports[`Social Showcase Links can handle more complex interactions 1`] = `
this is the first
<a href="//url.com"
   role="presentation"
>
  url.com
</a>
`;

exports[`Social Showcase Links should have a social editor 1`] = `
This is text
<a href="https://url.com"
   role="presentation"
>
  https://url.com
</a>
`;

exports[`Social Showcase Links should parse simple urls 1`] = `
<a href="//url.com"
   role="presentation"
>
  url.com
</a>
`;

exports[`Social Showcase Links should parse simple urls 2`] = `
<a href="//url.co"
   role="presentation"
>
  url.co
</a>
`;

exports[`Social Showcase Links should parse simple urls 3`] = `
<a href="//url.co.uk"
   role="presentation"
>
  url.co.uk
</a>
`;

exports[`Social Showcase Mentions @ should not trap the arrow keys 1`] = `
<p class>
  12
  <a role="presentation"
     href="/a"
     class="mention mention-at"
     data-mention-id="a"
     data-mention-name="at"
  >
    @a
  </a>
</p>
`;
